frontend/web/src/app/components/material/material-edit/material-edit.component.ts

```diff
@@ -0,0 +1,256 @@
+    const materialGrades: MaterialGrade[] = new Array<MaterialGrade>();
+    const materialProperties = {};
+    this.materialGradeFieldsForm.forEach((formGroup) => {
+      materialGrades.push(new MaterialGrade({
+        amount: formGroup.value.amount,
+        reusedTimes: formGroup.value.reusedTimes,
+      }));
+    });
+    material.materialGrades = materialGrades;
+    this.materialPropertiesFieldsForm.forEach((formGroup) => {
+      const v = formGroup.value.value;
+      const p = formGroup.value.property;
+      materialProperties[v] = p;
+    });
+    material.materialProperties = materialProperties;
+    if (this.creating) { // a new product
+      this.materialService.createMaterial(material).subscribe(
+        (data) => {
+          if (this.nav) {
+            this.back();
+          }
+          this.changed.emit(data);
+        }, (error) => {
+          console.log(error);
+          this.errorService.showAlert(error.verobose_message_header, error.verbose_message);
+        },
+      );
+    } else {
+      this.materialService.updateMaterial(material).subscribe(
+        (data) => {
+          if (this.nav) {
+            this.back();
+          }
+          this.changed.emit(data);
+        }, (error) => {
+          console.log(error);
+          this.errorService.showAlert(error.verobose_message_header, error.verbose_message);
+        },
+      );
+    }
+  }
+
+  createNewDigitalPart() {
+    this.router.navigate(['digital-parts', 'create']);
+  }
+
+  addMaterialGrade() {
+    const fields = {
+      reusedTimes: ['',
+        Validators.compose([Validators.required])],
+      amount: ['',
+        Validators.compose([Validators.required])],
+    };
+    this.materialGradeFieldsForm.push(this.formBuilder.group(fields));
+  }
+
+  deleteMaterialGrade(option) {
+    this.materialGradeFieldsForm.splice(option, 1);
+  }
+
+  addMaterialProperty() {
+    const fields = {
+      value: ['',
+        Validators.compose([Validators.required])],
+      property: ['',
+        Validators.compose([Validators.required])],
+    };
+    this.materialPropertiesFieldsForm.push(this.formBuilder.group(fields));
+  }
+
+  deleteMaterialProperty(option) {
+    this.materialPropertiesFieldsForm.splice(option, 1);
+  }
+
+  cancel() {
+    if (this.nav) {
+      this.back();
+    }
+    this.changed.emit(this.material);
+  }
+
+  back() {
+    this._location.back();
+  }
+
+  materialGradesValid(): boolean {
+    let valid = true;
+    if (this.materialGradeFieldsForm.length === 0) {
+      return false;
+    }
+    this.materialGradeFieldsForm.forEach((formGroup) => {
+      if (!formGroup.valid) {
+        valid = false;
+        return valid;
+      }
+    });
+    return valid;
+  }
+
+  materialPropertiesValid(): boolean {
+    let valid = true;
+    if (this.materialPropertiesFieldsForm.length === 0) {
+      return false;
+    }
+    this.materialPropertiesFieldsForm.forEach((formGroup) => {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Update this function so that its implementation is not identical to the one on line 230. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=typescript%3AS4144)